### PR TITLE
[SPARK-22640][PYSPARK][YARN]switch python exec on executor side

### PR DESCRIPTION
**Pull request by yaooqinn**
## What changes were proposed in this pull request?
```
PYSPARK_PYTHON=~/anaconda3/envs/py3/bin/python \
bin/spark-submit --master yarn --deploy-mode client \ 
--archives ~/anaconda3/envs/py3.zip \
--conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=py3.zip/py3/bin/python \
--conf spark.executorEnv.PYSPARK_PYTHON=py3.zip/py3/bin/python  \
/home/hadoop/data/apache-spark/spark-2.1.2-bin-hadoop2.7/examples/src/main/python/mllib/correlations_example.py
```
In the case above, I created a python environment, delivered it via `--arichives`, then visited it on Executor Node via `spark.executorEnv.PYSPARK_PYTHON`.
But Executor seemed to use `PYSPARK_PYTHON=~/anaconda3/envs/py3/bin/python` instead of `spark.executorEnv.PYSPARK_PYTHON=py3.zip/py3/bin/python`, then application end with ioe.

this pr aim to switch the python exec when user specifies it.

## How was this patch tested?

manually verified with the case above.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala [hunk 1]
```diff
@@ -57,9 +57,14 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
 
   require(funcs.length == argOffsets.length, "argOffsets should have the same length as funcs")
 
+  private val env = SparkEnv.get
+  private val conf = env.conf
+
   // All the Python functions should have the same exec, version and envvars.
   protected val envVars = funcs.head.funcs.head.envVars
-  protected val pythonExec = funcs.head.funcs.head.pythonExec
+  protected val pythonExec = conf.getOption("spark.executorEnv.PYSPARK_DRIVER_PYTHON")
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala]: Shouldn't this be handled in `deploy/PythonRunner`?

**Comment by jiangxb1987** [comment]
cc @cloud-fan @ueshin

File: core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala (hunk 1 above)
Comment (ueshin) [reply core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala]: Yeah, `PYSPARK_DRIVER_PYTHON` should be handled in `deploy/PythonRunner`.
This `PythonRunner` is for executors.

**Comment by ueshin** [comment]
Should we set the `pythonExec` during the initialization of `SparkContext` at [context.py#L191](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L191) instead of in `PythonRunner`? @jiangxb1987 @cloud-fan 

Btw, just for confirmation, is `spark.yarn.appMasterEnv.PYSPARK_PYTHON` working? @yaooqinn

**Comment by yaooqinn** [comment]
@ueshin cluster mode working, client not

**Comment by yaooqinn** [comment]
i guess specifing `PYSPARK_PYTHON=~/anaconda3/envs/py3/bin/python` overwrites spark.executorEnv.PYSPARK_PYTHON by [context.py#L156](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L156), where should use `setIfMissing`

**Comment by jerryshao** [comment]
I think in YARN we have several different ways to set `PYSPARK_PYTHON`, I guess your issue is that which one should take priority?

Can you please:

1. Define a consistent ordering for such envs, which one should take priority (spark.yarn.appMasterEnv.XXX or XXX), and document it.
2. Check if it works as expected for `spark.yarn.appMasterEnv.PYSPARK_PYTHON` and `spark.executorEnv.PYSPARK_PYTHON`.

I guess for other envs, it will also suffer from this problem, am I right?

**Comment by yaooqinn** [comment]
wx20171201-145631 2x

@jerryshao ENVs are specified ok by yarn, but the `pythonExec` is generated in `PythonRDD` at driver side and delivered to executor side within a closure

**Comment by jerryshao** [comment]
Oh, I see. You're running in client mode. So this one `--conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=py3.zip/py3/bin/python` is useless.

So I guess the behavior is expected. Because driver will honor `PYSPARK_PYTHON` env and ship it to executors. So the cluster will use same python executables.

With your above test, `/path/to/python` is different for driver and executors, will it bring in issues? Driver uses `PYSPARK_PYTHON` and executors uses `spark.executorEnv.PYSPARK_PYTHON` which points to different paths.

Normally I think we don't need to set PYSPARK_PYTHON in executor side.

Please correct me if I'm wrong.

**Comment by yaooqinn** [comment]
Yes, you are right. we should use same python executables. But the **same** might mean binary same not just same path

**Comment by jerryshao** [comment]
I'm a little concerned about such changes, this may be misconfigured to introduce the discrepancy between driver python and executor python, at least we should honor this configuration "spark.executorEnv.PYSPARK_PYTHON" unless the executables of `PYSPARK_PYTHON` sent by driver is not existed. (Just my two cents)

ping @zjffdu , any thought?

**Comment by vanzin** [comment]
Instead of setting `PYSPARK_PYTHON=~/anaconda3/envs/py3/bin/python`, what happens if you set `PYSPARK_DRIVER_PYTHON=~/anaconda3/envs/py3/bin/python`?

**Comment by yaooqinn** [comment]
@vanzin PYSPARK_DRIVER_PYTHON won't work because [context.py#L191](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L191) does't deal with it

**Comment by yaooqinn** [comment]
#### use spark-2.2.0-bin-hadoop2.7 numpy examples/src/main/python/mllib/correlations_example.py
### case 1 
|key|value|
|---|---|
|**PYSPARK_DRIVER_PYTHON**|~/anaconda3/envs/py3/bin/python|
|deploy-mode|**client**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv.PYSPARK_PYTHON| py3.zip/py3/bin/python |
|failure|Exception: Python in worker has different version 2.7 than that in driver 3.6, PySpark cannot run with different minor versions.Please check environment variables PYSPARK_PYTHON and PYSPARK_DRIVER_PYTHON are correctly set.|

### case 2

|key|value|
|---|---|
|**PYSPARK_DRIVER_PYTHON**|~/anaconda3/envs/py3/bin/python|
|deploy-mode|**cluster**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv.PYSPARK_PYTHON| py3.zip/py3/bin/python |
|failure|java.io.IOException: Cannot run program "/home/hadoop/anaconda3/envs/py3/bin/python": error=2, No such file or directory at org.apache.spark.**deploy.PythonRunner**$.main(PythonRunner.scala:91)|

### case 3 & 4

|key|value|
|---|---|
|**PYSPARK_DRIVER_PYTHON**|~/anaconda3/envs/py3/bin/python|
|deploy-mode|**cluster(3) client (4)**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_DRIVER_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv. PYSPARK_DRIVER_PYTHON | py3.zip/py3/bin/python |
|failure|Exception: Python in worker has different version 2.7 than that in driver 3.6, PySpark cannot run with different minor versions.Please check environment variables PYSPARK_PYTHON and PYSPARK_DRIVER_PYTHON are correctly set.|

### case 5 && 6 

|key|value|
|---|---|
|**PYSPARK_PYTHON**|~/anaconda3/envs/py3/bin/python|
|deploy-mode|**cluster(6)**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_DRIVER_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv. PYSPARK_DRIVER_PYTHON | py3.zip/py3/bin/python |
|failure|java.io.IOException: Cannot run program "/home/hadoop/anaconda3/envs/py3/bin/python": error=2, No such file or directory [**executor side PythonRunner**]|

### case 7
|key|value|
|---|---|
|**PYSPARK_PYTHON**|~/anaconda3/envs/py3/bin/python|
|deploy-mode|**cluster**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv. PYSPARK_DRIVER_PYTHON | py3.zip/py3/bin/python |
|**success **|--|

### case 8
|key|value|
|---|---|
|**PYSPARK_PYTHON**|~/anaconda3/envs/py3/bin/python|
|deploy-mode|**cluster**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv. PYSPARK_DRIVER_PYTHON | py3.zip/py3/bin/python |
|failure|java.io.IOException: Cannot run program "/home/hadoop/anaconda3/envs/py3/bin/python": error=2, No such file or directory [**executor side PythonRunner**]|


### case 9
|key|value|
|---|---|
|not setting~~PYSPARK_[DRIVER]_PYTHON~~|<empty>|
|deploy-mode|**client**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv. PYSPARK_DRIVER_PYTHON | py3.zip/py3/bin/python |
|failure|ImportError: No module named numpy|

### case 10
|key|value|
|---|---|
|not setting~~PYSPARK_[DRIVER]_PYTHON~~|<empty>|
|deploy-mode|**cluster**|
|--archives |~/anaconda3/envs/py3.zip|
|spark.yarn.appMasterEnv.**PYSPARK_PYTHON**|py3.zip/py3/bin/python|
|spark.executorEnv. PYSPARK_DRIVER_PYTHON | py3.zip/py3/bin/python |
|**success**| -- |


### my humble opinions
1. spark.executorEnv. PYSPARK_* takes no affect on executor side pythonExec, which is determined by driver.
2. if PYSPARK_PYTHON is specified then **spark.yarn.appMasterEnv.** should be suffixed  by **PYSPARK_PYTHON** not ~~PYSPARK_DRIVER_PYTHON~~
3. specifying PYSPARK_DRIVER_PYTHON fails all the cases, it may be caused by  https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L191 only deal with PYSPARK_PYTHON

**Comment by ueshin** [comment]
@yaooqinn What's the difference between case 7 and 8? Looks like the same configuration but the different result?

**Comment by yaooqinn** [comment]
@ueshin case 8 should be client deploy mode, excuse me for copy mistake, fixed

**Comment by ueshin** [comment]
@yaooqinn OK, I see the situation.

In client mode, I think we can't use `spark.yarn.appMasterEnv.XXX` which is for cluster mode. So we should use environment variable `PYSPARK_PYTHON` or `PYSPARK_DRIVER_PYTHON`, or corresponding spark conf, `spark.pyspark.python`, `spark.pyspark.driver.python`.

In cluster mode, we can use `spark.yarn.appMasterEnv.XXX` and if there exist `spark.yarn.appMasterEnv.PYSPARK_PYTHON` or `spark.yarn.appMasterEnv.PYSPARK_DRIVER_PYTHON`, they overwrite original environment variables.

Btw, `PYSPARK_DRIVER_PYTHON` is for only Driver, not Executors, so we should handle only `PYSPARK_PYTHON` in executor and the priority of `PYSPARK_DRIVER_PYTHON` is higher than `PYSPARK_PYTHON` in Driver.

Currently we handle only environment varibale but not `spark.executorEnv.PYSPARK_PYTHON` for executor so we should handle it at `api/python/PythonRunner` as you do now or [context.py#L191](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L191).

**Comment by yaooqinn** [comment]
I can `spark.executorEnv.PYSPARK_PYTHON` in `sparkConf` at executor side ,  because it is set at [context.py#L156](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L156) by [conf.py#L153](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/conf.py#L153)

**Comment by ueshin** [comment]
@yaooqinn I meant it is not used for `pythonExec`.

**Comment by yaooqinn** [comment]
@ueshin i see.

**Comment by yaooqinn** [comment]
@ueshin [context.py#L191](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L191) set for both driver and executor?

**Comment by ueshin** [comment]
@yaooqinn It is used for executors.

**Comment by vanzin** [comment]
I'm trying to understand what is https://github.com/apache/spark/blob/master/python/pyspark/context.py#L191 really achieving. It seems pretty broken to me and feels like the whole `pythonExec` tracking in the various places should be removed.

It causes this problem because it forces the executor to use the driver's python even if it's been set to a different path by the user.

It uses `python` instead of `sys.executable` as the default value.

And it ignores the `spark.pyspark.python` config value if it's set.

Instead, shouldn't the logic at https://github.com/apache/spark/blob/master/launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java#L304 be used in `PythonRunner` (except for the driver python config) to find out the executor's python to use?

**Comment by yaooqinn** [comment]
@vanzin 
according to @ueshin 's explanation, `PYSPARK_DRIVER_PYTHON` is only for driver, if executor follows the order of [SparkSubmitCommandBuilder.java#L304](https://github.com/apache/spark/blob/master/launcher/src/main/java/org/apache/spark/launcher/SparkSubmitCommandBuilder.java#L304), we may not need so many configs, right?

**Comment by vanzin** [comment]
That's what I said in my comment ("except for the driver python config").

**Comment by vanzin** [comment]
So, any updates here?

**Comment by vanzin** [comment]
@yaooqinn do you plan to update this PR?

**Comment by yaooqinn** [comment]
@vanzin I am not very familiar with python part [context.py#L191](https://github.com/yaooqinn/spark/blob/8ff5663fe9a32eae79c8ee6bc310409170a8da64/python/pyspark/context.py#L191), so handle it at `api/python/PythonRunner` as I did in this pr. 

Maybe someone else could help, sorry for the delay.

**Comment by tgravescs** [comment]
I didn't read the entire thread here but what you want is this:

 --archives hdfs:///python36/python36.tgz#python36 --conf spark.pyspark.python=./python36/bin/python3.6 --conf spark.executorEnv.LD_LIBRARY_PATH=./python36/lib --driver-library-path /opt/python36/lib --conf spark.pyspark.driver.python=/opt/python36/bin/python3.6

**Comment by vanzin** [comment]
@yaooqinn we should probably close this if you're not planning to look at the root of the problem, which seems to be on the python side.

**Pull request closed without merging**